### PR TITLE
Config file formalization and credential sets: part 1, Config Class

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This is part 1 of #3709.  It is part of a series to address #3661.

This PR implements a class, Inspec::Config, which represents the config file.  It is responsible for many tasks that were formerly in BaseCLI:

 * Providing per-command defaults
 * Merging defaults, config file, and CLI options
 * Parsing and validating formatter options
 * Performing other miscellaneous option finalization

Much of this class's code was copied from lib/inspec/base_cli.rb .  It will be removed in part 2 of the PR series.

This PR has a no prerequisites and may be merged directly to master, or to a rollup branch.

### CODE REVIEW

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb:56]: Are there tests for the paths through this? Also, should we validate that the `cfg_io` passed is a `StringIO`?

NOTE: I'm also curious why not to use a guard clause. I suspect this is a preference thing, if it is could you help me learn, offline or here, your choice.

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb]: Is there an issue to track this? This feels like there is more context here than this TODO line indicates.

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb]: Is STDIN being a TTY required? I'm wondering if this would affect users using SSH commands as a job runner (which may or not require a TTY).

```
ssh somehost 'inspec exec someprofile --config -' <<-EOF
{
  INSERT JSON HERE THAT MAY USE ENV VARS ON SSH HOST
}
EOF
```

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb:118]: I'm confused why this is a `!` method, it doesn't seem to modify any state.

File: test/unit/config_test.rb
Comment (jerryaldrichiii) [inline test/unit/config_test.rb]: Took me a while to understand where `'version'` came from. Could we change it to:

```
ex.message.must_match /Config was:.*version.*$/
```

Not necessary, but would have made it easier for me at a glance.

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb:128]: What's the reason for the `sort` here?

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb]: I'm curious how the schema will be validated in future PRs, right now I don't see a way to avoid

```
reporter: "json"
```

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb:210]: There's a lot going on here, I think this is covered end-to-end in your unit tests, but I'd like to see this method isolated and tests created for it for the sake of regression prevention.

I also feel you could split this into a few methods based on Hash or Array which might make it a bit easier to digest (and avoid having to disable the AbcSize cop)

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb]: This one goes a bit back, context can be found here:

https://github.com/inspec/inspec/pull/1918

https://github.com/inspec/inspec/pull/1918/files#diff-2b20044866abd29a7a781f5760238403R30

I think we should link it or summarize it here.

File: test/unit/config_test.rb
Comment (jerryaldrichiii) [inline test/unit/config_test.rb]: Dubious how so? Could you elaborate in the comment?

File: test/unit/config_test.rb
Comment (jerryaldrichiii) [inline test/unit/config_test.rb:252]: See other comment about context and 

https://github.com/inspec/inspec/pull/1918/files#diff-2b20044866abd29a7a781f5760238403R30

File: test/unit/config_test.rb
Comment (jerryaldrichiii) [inline test/unit/config_test.rb:1]: I'm holding off on the `chefstyle/rubocop -a` PR until the final product, but am still willing to do the work premerge as it takes me very little time. I'll avoid any style issues in this PR for this reason.

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb:101]: I don't see where `file_version` or `legacy_file` are used anywhere else. 
Any reason we shouldn't remove those methods and change this to?:

```
# Assume legacy file if no version is specified
if !@cfg_file_contents['version']
```

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb:38]: `pp` is fine here, but I'm wonder if the work you did with UI would prevent the need for it.

**Review by jerryaldrichiii — COMMENTED** [review]
Gave this a look through, but I'm hesitant to give this an Approve/Reject.

Specifically, I'm having trouble following the split context across PRs and am not sure I can hold the whole plan/direction in my head. I might need to look at the functional tests/docs changes when they are added to help understand it. With that in mind, I would like us use a feature branch and then have a final review of the end product.

Of course, my trepidation might also resolved when we do a group review.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:118]: `!` is also often used to mark methods that are expected to generate exceptions.

File: test/unit/config_test.rb
Comment (clintoncwolfe) [reply test/unit/config_test.rb]: When we read a JSON file, should we expect the String "true" to be returned as a String, or as `true`, an instance of TrueClass?

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:38]: There is nothing in the Inspec::UI work that deals with pretty-printing deep structures.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb]: No, I just personally always felt that `--json-config` was a terrible name for 'the option that tells you where the config file is'.  

I can add it to #3661.

Note that `--json-config` is still supported, it just isn't documented. We could make that more explicit.

Also note that any comments like `TODO: deprecate...` are intended to be replaced by proper deprecations once #3690 lands.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:56]: We don't expect cfg_io to be a StringIO - it can be anything IO-like. In actual use that's likely to be an IO::File, while in testing it is easier to use a StringIO.  If there is no file, then we use a StringIO to treat it in the same way.

Comment (clintoncwolfe) [reply lib/inspec/config.rb:56]: Guard clauses are used to short-circuit logic, usually a break, next, or return.  Use them you have things like `return true if condition...`.

In this case, there is a lot of logic on either side of the unless.  That's not a situation for a guard clause.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb]: As mentioned in the description, most of the code was literally copied from base_cli.rb . I don't vouch for the logic of checking for a TTY, but it is how inspec currently behaves:

https://github.com/inspec/inspec/blob/feefdcf9afba80f4e666e6dad5ffce6bd30019dd/lib/inspec/base_cli.rb#L359

File: test/unit/config_test.rb
Comment (clintoncwolfe) [reply test/unit/config_test.rb]: I greatly prefer to check for individual elements of an UX message on an element-by-element basis. Rather than altering the test to only match on one regex (that could fail for any number of reasons, and will only grow in complexity over time), checking for one thing at a time make it obvious which piece is missing.

A comment with example output would be informative though - I'll add that.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:128]: Honestly not sure!  I think so the user sees a nice sorted list in the 'valid entries' in the error message? I could certainly sort in source.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb]: I'm not sure I follow your concern.  There is no intended change to the logic of parsing or validating the reporters.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:210]: This is pretty much directly copied out of base_cli.rb. I don't claim to make any changes to the code, nor improve the testing on it.

https://github.com/inspec/inspec/blob/feefdcf9afba80f4e666e6dad5ffce6bd30019dd/lib/inspec/base_cli.rb#L136

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:101]: A big point of this PR series is to formalize the config file format, and get it into a shape where can version the format, and validate it.

Having the Config object be able to report on the detected version is thus key to the intent of the PR.

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [reply lib/inspec/config.rb:56]: I don't guess I'm following, I don't see any logic surrounding unless, couldn't the first line be:

```
return cfg_io if cfg_io
```

Comment (jerryaldrichiii) [reply lib/inspec/config.rb:56]: I do think we need some sort of type checking on `cfg_io` as well too, otherwise anything passed into `initialize`, that doesn't respond to `.read` will fail here: 

https://github.com/inspec/inspec/blob/09e0579f5c34b642de939dbfb54aaf220249bf03/lib/inspec/config.rb#L76

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [reply lib/inspec/config.rb:128]: Eh, that's a lot of load on the developer `.sort` is fine here, CPU cycles are cheap :smile:

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [reply lib/inspec/config.rb:210]: Ah, I was thinking about the Boy Scout Rule you've mentioned in the past. Does it only apply to small things (I recognize adding tests for this would be significant work...at least for me)

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [reply lib/inspec/config.rb]: I was concerned if someone had an invalid config (though valid JSON). For example if someone used just `reporter: "json"` then https://github.com/inspec/inspec/blob/09e0579f5c34b642de939dbfb54aaf220249bf03/lib/inspec/config.rb#L147 would fail with a stacktrace.

File: test/unit/config_test.rb
Comment (jerryaldrichiii) [reply test/unit/config_test.rb]: Ah, could we elaborate in the comment?

File: lib/inspec/config.rb
Comment (jerryaldrichiii) [inline lib/inspec/config.rb]: Given https://github.com/inspec/inspec/pull/3710/files/09e0579f5c34b642de939dbfb54aaf220249bf03#r248157588 shouldn't this method also be a `!` method?

File: test/unit/config_test.rb
Comment (clintoncwolfe) [reply test/unit/config_test.rb:252]: This is just copied from the base_cli_test.rb

https://github.com/inspec/inspec/blob/feefdcf9afba80f4e666e6dad5ffce6bd30019dd/test/unit/base_cli_test.rb#L11

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb]: Good idea!  This was just copied from base_cli.rb, and didn't follow that pattern.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:56]: I personally don't see that (passing a non-IO object) as likely - this is an internal InSpec object - but it's no big deal to add a duck check.

Comment (clintoncwolfe) [reply lib/inspec/config.rb:56]: And regarding the guard clause: you are totally right, I just didn't read down far enough.  This is an ideal case for a guard clause.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb]: So, when Inspec::Config finds a problem, it throws an exception.  Things that consume Inspec::Config (none of which are included on this PR) are then responsible for catching exceptions and handling them - which might mean reporting a graceful error to the user, with no stacktrace.

The existing code throws exceptions, as well - see https://github.com/inspec/inspec/blob/feefdcf9afba80f4e666e6dad5ffce6bd30019dd/lib/inspec/base_cli.rb#L171 

The existing code calls validate_reporters through a long chain, eventually surfacing as the `opts` call which is used as one of the first things in nearly every command method.  And each command method provides a facility for handling exceptions - for example, here is `inspec exec`'s exception handler block:

https://github.com/inspec/inspec/blob/master/lib/inspec/cli.rb#L252

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:210]: It's a rule of thumb, I guess, and relies on intuition.

The goal of this PR series is to get the config reading and validation into a proper class, and then expose a simpler way of reading credentials. So, I must copy a lot of code from where it was (scattered as class methods in Inspec::BaseCLI) to its new home. (On the 2nd PR of the series, I delete it from its previous location - which perhaps I should have done on this PR). In order to help people follow the changes, I'm just doing simple copying as much as possible. Making improvements (like adding tests, renaming vars, etc) would add noise to the PR.

We could improve testing of the reporter parser on a later PR, either on this series, or treat that as a separate issue - I think that might take some real effort to really exercise that facility.  And the problem with adding tests, you know... you may find bugs, which then you must fix! Which makes me thing adding tests for the reporter parser is probably best as a separate issue.

File: test/unit/config_test.rb
Comment (robbkidd) [inline test/unit/config_test.rb:72]: I appreciate your example data. 

I see you.

File: lib/inspec/config.rb
Comment (marcparadise) [inline lib/inspec/config.rb:70]: Is reading config from stdin a new behavior, or something that was previously supported?

File: lib/inspec/config.rb
Comment (marcparadise) [inline lib/inspec/config.rb:165]: Can we move this into the loop at line 153, and avoid the second loop?

File: lib/inspec/config.rb
Comment (marcparadise) [inline lib/inspec/config.rb:182]: Making sure I understand this - this was due to supporting it in previous versions in different locations/'cli', but it's now required under `reporting` at the top level?

File: lib/inspec/config.rb
Comment (marcparadise) [inline lib/inspec/config.rb:211]: To simplify testing, would it be worthwhile to break this into two separate impl methods, `finalize_parse_reporters_array` and `finalize_parse_reporters_hash`? I know that it's existing code, but this might be a good opportunity to clean it a bit (unless that's planned/done in subsequent PRs?)

**Review by marcparadise — CHANGES_REQUESTED** [review]
This looks like it will make config/cli handling quite a bit easier to reason about.   A few questions and a couple of comments inline.

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb:70]: It was previously supported - most of this is copied from lib/inspec/base_cli.rb .

**Comment by clintoncwolfe** [comment]
As this PR breakout is confusing everyone, I'm withdrawing this and will re-open in a different way.

**Pull request closed without merging**
